src/main/java/com/crafting/Knowledge.java

```diff
@@ -0,0 +1,45 @@
+package com.crafting;
+
+import com.minecolonies.util.InventoryUtils;
+import net.minecraft.inventory.InventoryCrafting;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipe;
+import net.minecraft.world.World;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+
+/**
+ * One bit of knowledge from the crafting system.
+ */
+public class Knowledge
+{
+    @NotNull
+    private final IRecipe         recipe;
+    @NotNull
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Perhaps 'requirements' could be replaced by a local variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ASingularField)